### DISCUSSION TITLE
The x509 certificate fails to be parsed when wolfssl is used to replace openssl.

### DISCUSSION THREAD

**Discussion by shenhaicheng** [strongswan/strongswan#1672] [Q&A] [answered]
We plan to enable the wolfSSL encryption crypto plugin to replace openSSL. The new configuration options are:
**--enable-wolfssl --with-crypto=wolfssl --enable-x509 --enable-pem** 

When x509 certificate encryption is used to link to an ipsec VPN, certificate decryption fails.
The log is as follows:
**loading '/var/vpn/swanctl/x509/rpub1.pem' failed: parsing X509 certificate failed**

Does the wolfSSL plugin support the x509 certificate link? If supported, how should we configure.

Thanks in advance~

**Comment by tobiasbrunner** [comment]
> --with-crypto=wolfssl

There is no such option (you should get a corresponding warning message).

Since you haven't posted the complete configure options or a detailed log, we can only guess. But you'll definitely also need the _pkcs1_ plugin to load certificates (and possibly the _pkcs8_ plugin if you also want to load private keys in PKCS#8 format).

**Reply by shenhaicheng** [reply]
Thanks for your answer.  We add the _pkcs1_ plugin, but the certificate loading still fails. And then we rechecked strongswan.conf and found that **x509** was filled in as **X509**. The following is the modified strongswan.conf.

> swanctl {                                                                                                                           
>     load = pkcs1 constraints pubkey wolfssl random x509                                                                             
> }                                                                                                                                   
>                                                                                                                                     
> charon {                                                                                                                            
>     load = random nonce pem pubkey counters aes sha1 sha2 md5 hmac pkcs1 curve25519 wolfssl kernel-netlink socket-default updown vici x509                                                                                                                              
>          crypto {                                                                                                                   
>            backend = wolfssl                                                                                                        
>          }                                                                                                                          
>     filelog {                                                                                                           
>         charon-debug-log {                                                                                                          
>                 path = /var/log/charon_debug.log                                                                                    
>                 default = 4                                                                                                         
>                 net = 1                                                                                                             
>                 enc = 1                                                                                                             
>                 asn = 1                                                                                                             
>                 job = 1                                                                                                             
>                 ike_name = yes                                                                                                      
>                 append = no                                                                                                         
>                 flush_line = yes                                                                                                    
>             }                                                                                                                       
>     }                                                                                                                               
> }  

By using this configuration file, the problem is solved. But I have one more question I'd like to ask, when we use the _openssl_ plugin ,  the x509 certificate can be properly loaded. However, when we switch to _wolfssl_ plugin , we have to load the _x509_ plugin .Can I understand that _wolfssl_  plugin and _x509_ plugin  should be used together? Thanks!

**Reply by tobiasbrunner** [reply, marked as answer]
> And then we rechecked strongswan.conf and found that **x509** was filled in as **X509**.

We usually don't recommend using such `load` statements, see [here](https://docs.strongswan.org/docs/5.9/plugins/pluginLoad.html#_static_load_list).

> crypto {
> backend = wolfssl
> }

Like `--with-crypto=wolfssl` this is not a thing. Where did you get these from?

> However, when we switch to _wolfssl_ plugin , we have to load the _x509_ plugin .Can I understand that _wolfssl_ plugin and _x509_ plugin should be used together?

Yes, the _wolfssl_ plugin does not implement certificate parsing. Only the _x509_ and the _openssl_ plugins provide this feature (the former provides more features e.g. support for OCSP).

I assume it would theoretically be possible to use wolfSSL with the _openssl_ plugin if you used their OpenSSL compatibility layer. The _wolfssl_ plugin uses the lower-level wolfCrypt API directly.

**Comment by shenhaicheng** [comment]
Thanks for answering! 
We found some solutions on the Internet before, resulting in some incorrect configuration items. We'll revise it here.
